### ISSUE TITLE
npm5 npm outdated leads to crash with cb() never called

### ISSUE THREAD

**Issue by dananichev** [npm/npm#16825]
#### I'm opening this issue because:

  - [x] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [x] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?
```js
beam-web git:ISC-1001 ✗ ✶ ✩  ❯ npm outdated
npm ERR! cb() never called!

npm ERR! This is an error with npm itself. Please report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/dananichev/.npm/_logs/2017-05-31T12_57_45_180Z-debug.log
 ⌚︎ 7s
```

#### How can the CLI team reproduce the problem?
I'm not sure. Here is my debug-log https://gist.github.com/dananichev/86e10dcca3614cf8965fe03b621f8bdc

### supporting information:

 - `npm -v` prints: 5.0.0
 - `node -v` prints: v7.10.0
 - `npm config get registry` prints: http://builds.ic360:4873/
 - Windows, OS X/macOS, or Linux?: OS X
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

Also, i tried to clean my cache (both ``npm cache clean`` and ``sudo rm -rf ~/.npm``). And i'm using ``nodenv`` to run multiple Node versions.

**Comment by gizburdt** [comment]
Same error:

```
npm ERR! cb() never called!

npm ERR! This is an error with npm itself. Please report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/Gijs/.npm/_logs/2017-05-31T13_13_59_135Z-debug.log
```

npm: 5.0.0
node: 7.10.0

**Comment by gizburdt** [comment]
Updating Node to 8.0.0 fixed it for me

**Comment by dananichev** [comment]
@gizburdt sadly, i can't update all our servers to Node 8 -- too early for that. 
Thanks for useful info anyways.

**Comment by alex-kinokon** [comment]
I can reproduce this error even with Node 8.0.0

**Comment by dananichev** [comment]
After updating to ``npm@5.0.1`` issue is still reproducable.

**Comment by robin-marr-lh** [comment]
I'm still seeing this issue with npm 5.0.2 and node 8, here's a package.json that will reproduce it consistently:
https://gist.github.com/robin-marr-lh/7c2aab17249bd21a15f40cce7ecded47

**Comment by antipin** [comment]
I have the same issue. Noticed it happens only when I have a tarball in my deps section:
```
{
  "name": "...",
  "private": true,
  "version": "0.0.1",
  ...
  "dependencies": {
    ...
    "prop-types": "https://registry.npmjs.org/prop-types/-/prop-types-15.5.6.tgz",
    ...
  }
  ...
}
```

**Comment by JustinPierce** [comment]
I am having this issue, as well. Following @antpin's lead, I've found that removing the one dependency we have that points to a GitHub URL resolves the issue.

**Comment by max-winderbaum** [comment]
I can confirm that this happens specifically when `package.json` contains either github url or tarball dependencies. We have both and I tried removing one or the other to no avail. When I removed both it worked.

**Comment by jhchill666** [comment]
Same happens with a private npm module:

```
"dependencies": {
    "some-module": "git+ssh://git@foo.com:7999/sm/some-module.git#v0.0.1",
}
```

`node: v8.4.0`
`npm: v5.3.0`

**Comment by joshclow** [comment]
I've narrowed this down. The problem is in `shouldUpdate` in `outdated.js`. That code is expecting to call the old version of `cache.add` that took a callback (`onCacheAdd`), but the cache code is now returning a promise from `pacote.prefetch`. As a consequence, the process neither exits cleanly, nor calls into the main error handling callback (the "cb" from the error message).

Looking through the history, I've found the old `cache/add-remote-tarball.js` and `cache/add-remote-git.js` and am now trying to understand what data they returned so I can move the callback into `.then` and do the equivalent checks from there.
